dev/com.ibm.ws.org.apache.myfaces.2.3/bnd.bnd

```diff
@@ -84,58 +85,62 @@ Include-Resource: \
-  org.apache.myfaces.config.impl.digester.elements;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.flow;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.flow.cdi;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.lifecycle;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.push.cdi;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.renderkit;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.renderkit.html;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.shared.context.flash;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.shared.taglib;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.shared.taglib.core;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.shared.util;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.spi.impl;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.taglib.core;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.taglib.html;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.compiler;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.component;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.el;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.impl;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.pool;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.pool.impl;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.composite;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.jsf;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.jsf.core;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.jsf.html;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.jstl.core;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.jstl.fn;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.tag.ui;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.view.facelets.util;thread-context=true;mandatory:="thread-context", \
-  org.apache.myfaces.webapp;thread-context=true;mandatory:="thread-context"
+  org.apache.myfaces.*;version=${Implementation-Version}, \
```

Comment:
This export of org.apache.myfaces.* followed by a series of explicit org.apache.myfaces.XXX exports with thread-context=true is confusing.

Are there other org.apache.myfaces packages that we need to export that we do not want thread-context=true on which the `*` expansion covers?  Without building myself I am not sure what the resulting export package statement generated by BND is here.  Can you confirm if the resulting bundle manifest has the expected exports with `thread-context=true` for the appropriate packages.

Reply:
I agree that this is a little confusing.  Yes, the `*` expansion covers other packages that aren't intended to be `thread-context=true`.  My guess is that the `org.apache.myfaces.*` statement was put in to avoid needing to enumerate all of the 94 packages in this bundle.  

The resulting manifest _does_ have the correct `thread-context=true` exports listed, however those packages are duplicated, eg:
```
Export-Package:com.ibm.ws.jsf.cdi;version="1.0.0",
...
org.apache.myfaces.application;version="2.3.3",
...
org.apache.myfaces.application;thread-context=true;version="2.3.3",
...
```
Despite that duplication, this must provide the correct thread context behavior since the bundle has been working this way for some time.  Regardless, is this something we should change - ie. getting rid of the `*` and listing every package?

Reply:
OK, that is what I suspected would happen.  I think simply moving the wildcard `*` import to be after the explicit ones would get rid of the duplicates.  I think `*` matching will not export the package again if it is already exported, but the explicit export is always exported even if the package is already exported from a match before it.  That is what would happen for Import-Package at least, but perhaps the Export-Package behaves differently because a bundle is allowed to export a package multiple times with differing attributes.

Reply:
You're right, moving the wildcard export to the end of the Export-Package list gets rid of the duplicated packages.  I'll do a commit with that change.